src/bosh-director/lib/bosh/director/instance_updater/update_procedure.rb

```diff
@@ -0,0 +1,208 @@
+module Bosh::Director
+  class InstanceUpdater
+    class UpdateProcedure
+      attr_reader :instance, :instance_plan, :options, :instance_report, :action, :context, :links_already_bound_to_instance
+
+      def initialize(instance,
+                     instance_plan,
+                     options,
+                     blobstore,
+                     needs_recreate,
+                     instance_report,
+                     disk_manager,
+                     rendered_templates_persistor,
+                     vm_creator,
+                     links_manager,
+                     ip_provider,
+                     dns_state_updater,
+                     logger)
+        @instance = instance
+        @instance_plan = instance_plan
+        @options = options
+        @blobstore = blobstore
+        @needs_recreate = needs_recreate
+        @instance_report = instance_report
+        @disk_manager = disk_manager
+        @rendered_templates_persistor = rendered_templates_persistor
+        @vm_creator = vm_creator
+        @links_manager = links_manager
+        @ip_provider = ip_provider
+        @dns_state_updater = dns_state_updater
+        @logger = logger
+        @action = calculate_action
+        @context = calculate_context
+      end
+
+      def to_proc
+        -> { perform }
+      end
+
+      def perform
+        # Optimization to only update DNS if nothing else changed.
+        if dns_change_only?
+          @logger.debug('Only change is DNS configuration')
+          update_dns
+          return
+        end
+
+        unless instance_plan.already_detached?
+          handle_already_detached_instance_plan
+
+          # desired state
+          if instance.state == 'stopped'
+            # Command issued: `bosh stop`
+            instance.update_state
+            return
+          end
+        end
+
+        if instance.state == 'detached'
+          handle_detached_instance
```

Comment:
It looks like some of the logic in `handle_detached_instance` is repeated in `finish_up` (i.e. releasing network plans, links, dns, updating variable set, etc.). Is there a way we can only have this in one place?